### DISCUSSION TITLE
Extension methods for integration with ASP.NET Core services

### DISCUSSION THREAD

**Discussion by m3nax** [kubernetes-client/csharp#945] [Ideas]
I would like to implement extension methods for registering the kubernetes client through DI

Interface I plan to extend: **IServiceCollection**

Registration (example in net6.0 style):
``` CSHARP
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddKubernetesClient(kccb =>
{
   kccb.FromConfigFile();
   // other variants like
   // kccb.FromKubeconfigEviomentVariable();
   // kccb.FromInClusterConfig();

   kccb.UseContext("contextName"); 
   // other configurations
});
```

Usage (example in net6.0 ApiController style):
``` CSHARP
[Route("api/[controller]")]
[ApiController]
public class TestController : ControllerBase
{
  // Constructor injection
  public TestController(IKubernetes kubernetesClient)
  {
     //CODE
  }

  // WebMethod injection
  public IActionResult TestMethod([FromService] IKubernetes kubernetesClient)
  {
     //CODE
  }
}
```

Usage (example in net6.0 Minimal api style):
``` CSHARP
// service registration omitted

var app = builder.Build();

 // WebMethod injection
app.MapGet("/testmethod", (IKubernetes kubernetesClient) =>
{
     //CODE
})
```

usage (example in net6.0 Service Worker style):
``` CSHARP
public class Worker : BackgroundService
{
    private readonly ILogger<Worker> logger;
    private readonly IKubernetes kubernetesClient;

    public Worker(ILogger<Worker> logger, IKubernetes kubernetesClient)
    {
        this.logger = logger;
        this.kubernetesClient = kubernetesClient;
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        while (!stoppingToken.IsCancellationRequested)
        {
            logger.LogInformation("Worker running at: {time}", DateTimeOffset.Now);

            // CODE

            await Task.Delay(1000, stoppingToken);
        }
    }
}
```

Important points to discuss:
- kcb (kubernetes client configuration builder): divide the configuration from its construction and move the configuration validation to the builder class
- methods design and use cases in asp net core
- UseContext("name"): method for selecting which configuration to use
- other points that I have missed

Does this sound like a good idea?

**Comment by tg123** [comment]
sounds `AddSingleton(new k8sclient())` is enough to me
any reason why the ext is needed?

**Reply by m3nax** [reply]
In the future can be extended with named instance with different configuration, like HttpClientFactory
 https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests

Can be integrated with HttpClientFactory in order to recycle the httpclient used by the library, removing the need to register it as a sigleton and to preallocate the resources.

Kubernetes instance, for simple interrogations like GET or patch, can be created when requested.

**Reply by tg123** [reply]
k8s client does not allow customized httpclient due to watch timeout
see
https://github.com/kubernetes-client/csharp/blob/56646aeed3ee289146be9efda4d8da8756b8ac69/src/KubernetesClient/Kubernetes.ConfigInit.cs#L142

see #725
and 
the httpfactory was totally removed in #785


let me know if you have any idea to workaround the `timeout` otherwise, it is considered as a bug

**Reply by m3nax** [reply]
In the file [csharp/src/KubernetesClient/Kubernetes.ConfigInit.cs](https://github.com/kubernetes-client/csharp/blob/56646aeed3ee289146be9efda4d8da8756b8ac69/src/KubernetesClient/Kubernetes.ConfigInit.cs) 

How reproduce timeout error?

Here's what I tried to do:
- I changed the HttpClient initialization code as follow.
``` CSHARP

            HttpClient = new HttpClient(FirstMessageHandler, false)
            {
                //Timeout = Timeout.InfiniteTimeSpan,
            };

            Console.WriteLine($"Default timeout: {HttpClient.Timeout}");

```
- I added the Datetime.Now.ToString() to the log of the watch example
- I started the example program

Output:
```
Default timeout: 00:01:40
==on watch event==
13/07/2022 11:41:46 test-pod-744fdc5b-b5dkq - Added
==on watch event==
==on watch event==
13/07/2022 11:48:12 test-pod-77968dbf96-n4kls - Added
==on watch event==
```

After 7 minutes Watch program still works.
What i'm doing wrong?

**Reply by m3nax** [reply]
Commented this part of code and tested again:
``` CSHARP
           FirstMessageHandler = HttpClientHandler = new SocketsHttpHandler
            {
                //KeepAlivePingPolicy = HttpKeepAlivePingPolicy.WithActiveRequests,
                //KeepAlivePingDelay = TimeSpan.FromMinutes(3),
                //KeepAlivePingTimeout = TimeSpan.FromSeconds(30),
            };
```
Same result:
``` 
Default timeout: 00:01:40
==on watch event==
13/07/2022 12:03:21 test-pod-659445b4c5-b2fbq - Added
==on watch event==
==on watch event==
13/07/2022 12:13:55 test-pod-659445b4c5-b2fbq - Modified
==on watch event==
``` 
After 10 minutes without any changes Watch program still works.

**Reply by tg123** [reply]
testcase here
#165 
https://github.com/kubernetes-client/csharp/blob/85755ccb3ef069a875c8e278ced85ffdae99e07c/tests/KubernetesClient.Tests/WatchTests.cs#L583

**Reply by m3nax** [reply]
Thank
